### ISSUE TITLE
GPU & encoding problems with translate.py

### ISSUE THREAD

**Issue by shuoyangd** [OpenNMT/OpenNMT-py#70]
Hi All,

So I just trained my first model with OpenNMT-py, everything looks great and easy to play with. But when I started to run the inference part, things start to get tricky.

First, since I was translate from Chinese to English, everything in `translate.py` that's dealing with files and I/O is not working because of encoding problems. I fixed this by substituting `open` with `codecs.open`. You may want to fix this, or I can submit a pull request after more through testing.

(Not sure if this is relevant, but I'm using Python 3.6.1.)

Second, I assume when I pass `-gpu 0` I'm just using GPU device 0, but it does not look like so -- here is the `nvidia-smi` output when I run the program:

```
Tue Jun 20 17:38:50 2017
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 367.48                 Driver Version: 367.48                    |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|===============================+======================+======================|
|   0  Tesla K20m          On   | 0000:02:00.0     Off |                    0 |
| N/A   25C    P0    85W / 225W |    871MiB /  4742MiB |     93%   E. Process |
+-------------------------------+----------------------+----------------------+
|   1  Tesla K20m          On   | 0000:03:00.0     Off |                    0 |
| N/A   23C    P0    46W / 225W |   1300MiB /  4742MiB |      0%   E. Process |
+-------------------------------+----------------------+----------------------+

+-----------------------------------------------------------------------------+
| Processes:                                                       GPU Memory |
|  GPU       PID  Type  Process name                               Usage      |
|=============================================================================|
|    0     27148    C   python                                         869MiB |
|    1     27148    C   python                                        1298MiB |
+-----------------------------------------------------------------------------+
```

which seems to indicate there is another mysterious process which is taking a lot of memories but not doing anything. Any idea what started that process? And is there any chance I can get rid of it?

Here is the exact command I used to run `translate.py`:

```
python ~/nmt/opennmt/translate.py -gpu 0 -model $PWD/model/model_acc_58.17_ppl_7.80_e13.pt -src data/eval08.bpe.zh -tgt data/eval08.en.tok -output out
```

Thanks!

**Comment by shuoyangd** [comment]
I think I found the cause to the problem.

The serialized pytorch model seem to preserve the location where it was trained, so for example if the model was trained on device 0 and we want to translate on device 1, pytorch will end up loading the model to device 0 and translate on device 1, which is why we are using two GPUs here. Even worse, if we want to use this model to translate on a machine that has no GPU, it will crash for sure because pytorch won't able to find any device to load the model there.

I found a solution [here](https://discuss.pytorch.org/t/loading-weights-for-cpu-model-while-trained-on-gpu/1032). In short, in line 21 of [onmt/Tranlator.py](https://github.com/OpenNMT/OpenNMT-py/blob/master/onmt/Translator.py), instead of using

```
checkpoint = torch.load(opt.model)
```

this should have been used:

```
checkpoint = torch.load(opt.model, map_location=lambda storage, loc: storage)
```

which means we should load the model to whatever device we are using.

I made the change and tested on a multi-gpu machine as well as a cpu-only machine and it seems to fix the problem.
